net: ethernet: Clean an obscure ref/unref logic in ethernet_send()

ozhuraki opened the pull request:
Do not extra ref and then obscurely unref the packet inside the statistics update function.
    
Actually, this extra ref/unref isn't needed here at all.
    
The packet is unreferenced only on a successful send, statistics updating can be done before the unref in a clean and understandable way.

pfalcon commented:
It's weird that this didn't have reviewers assigned automatically.

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016 Intel Corporation.
+ * Copyright (c) 2016-2018 Intel Corporation.
```
That has nothing to do here

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -116,10 +116,10 @@ static inline void ethernet_update_length(struct net_if *iface,
 static void ethernet_update_rx_stats(struct net_if *iface,
 				     struct net_pkt *pkt, size_t length)
 {
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
```
The issue with this is that some compiler options will generate warnings on unused arguments. I don't see why you change that

ozhuraki replied:
I removed ifdefs around updating the error stat.

Take another look at the sucessful case, there's no need for dummy here. Or, am I missing anything?

tbursztyka replied:
I mean all #ifdef about CONFIG_NET_STATISTICS_ETHERNET can be removed here, since eth_stats_update_bytes_rx(), eth_stats_update_pkts_rx()  and so on are already provided as dummies in eth_stats.h
So compiler will optimize by itself when it will see empty statements in the if/else: it will remove this part entirely. (in case CONFIG_NET_STATISTICS_ETHERNET is unset)

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -566,17 +559,18 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
-		net_pkt_unref(pkt);
+
+	if (ret != 0) {
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
```
get rid of this #ifdef, eth_stats_update_errors_tx() has a dummy version when CONFIG_NET_STATISTICS_ETHERNET is unset

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -566,17 +559,18 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+		eth_stats_update_errors_tx(iface);
+#endif
+		goto error;
```
You don't need the goto, just have an else in case ret is 0. 
And set ret to net_pkt_get_len().

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -566,17 +559,18 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
-
 	ret = api->send(net_if_get_device(iface), pkt);
-	if (!ret) {
-		ret = net_pkt_get_len(pkt);
```
ethernet_send() needs to return the length sent on success

ozhuraki replied:
OK, fixed.

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -498,15 +498,15 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 	struct ethernet_context *ctx = net_if_l2_data(iface);
 	u16_t ptype;
-	int ret;
+	int err;
```
See my comment in 2nd patch, ret needs to be set as pkt length on succes so this variable can return either error status or successful length sent. Keep it as ret.

tbursztyka reviewed (changes requested):
2nd patch is worth keeping, just apply the comment.

ozhuraki commented:
@tbursztyka, @pfalcon 
ping

tbursztyka reviewed (changes requested):
Why merging the 2 first patches? they are not related. Anyway, the original 1st could be kept, see my new comment.

And apply the other comments I made.

ozhuraki committed:
net: ethernet: Clean an obscure ref/unref logic in ethernet_send()

Do not extra ref and then obscurely unref the packet inside
the statistics update function.

Actually, this extra ref/unref isn't needed here at all.

The packet is unreferenced only on a successful send, statistics
updating can be done before the unref in a clean and understandable way.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@tbursztyka 

> Why merging the 2 first patches? they are not related. Anyway, the original 1st could be kept, see my new comment.

There's no point in the first one after the changes.

> And apply the other comments I made.

The length is updated, regarding ifdefs on successful send, please take another look.

jukkar commented:
recheck

jukkar commented:
@nashif shippable seems to be stuck for this PR, can you do something to it?
